cont.c

```diff
@@ -2348,10 +2352,19 @@ rb_fiber_m_resume(int argc, VALUE *argv, VALUE fiber)
  *  blocks.
  */
 static VALUE
-rb_fiber_raise(int argc, VALUE *argv, VALUE fiber)
+rb_fiber_raise(int argc, VALUE *argv, VALUE fiber_value)
 {
+    rb_fiber_t *fiber = fiber_ptr(fiber_value);
     VALUE exc = rb_make_exception(argc, argv);
-    return rb_fiber_resume_kw(fiber, -1, &exc, RB_NO_KEYWORDS);
+    if (RTEST(fiber->resuming_fiber)) {
```

Comment:
What does this state represent? A Fiber suspended in `other_fiber = Fiber.new { sleep }; f = Fiber.new { other_fiber.resume }.resume`?
I think we could raise in that Fiber in that case too, so then Fiber#raise always works and just raises the exception wherever the Fiber is suspended.

I think we should still forbid to Fiber#raise on a currently-running Fiber (which is either Fiber.current, or the current Fiber of another Thread). Not a big deal for Fiber.current but it just seems weird to not use plain `raise` in that case.

Reply:
n.b. All I'm doing here is giving a different error message. Otherwise a `FiberError` would still have been raised inside either `rb_fiber_resume_kw` or `rb_fiber_transfer_kw` (which I extracted from `rb_fiber_m_transfer`).

The very first protection of https://bugs.ruby-lang.org/issues/17221 makes it impossible to resume or transfer into a resuming fiber. I strongly agree with all four of the new protections in that ticket and think they should be adhered to absolutely by all fiber switches (although it does break a trampoline trick I've been using since ruby 2.4).

Reply:
Although maybe now you see one of the things that is special about my `Fiber#cancel`PR #3766? :wink: 

#3766 strictly adheres to the all of the protections in https://bugs.ruby-lang.org/issues/17221 while still allowing any and every fiber to be canceled.  By propagating the cancellation down to `fiber->resuming` it eventually hits the last fiber in the chain, which will eventually yield back up the chain until the originally canceled fiber is reached.

Reply:
I don't see the difference, Fiber#cancel is conceptually just Fiber#raise except it's an un-rescue-able exception instead of a regular Ruby exception, no?
The fact rb_fiber_resume_kw/rb_fiber_transfer_kw are used internally for Fiber#raise just seems implementation details to me.

#3766 follows the `resuming_fiber` to find a non-resuming Fiber, I see. Maybe we can do exactly the same with Fiber#raise, any issue with that?

Reply:
Is `return` just an unrescuable ruby exception? Is `break` just an unrescuable ruby exception? :wink:

Cancellation is *not* a special case of error handling, and I don't know why we'd want to try to tunnel these forms of early exit through exceptions. These are two very different semantic concepts with two different sets of concerns and they should follow different rules. Ruby considers this such an important and distinct concept that we have *two* keywords *and* a Kernel method for early exit vs only one Kernel method for raising exceptions. :wink:

To me, `raise` means "I don't know how to handle this, but maybe some stack frame above me does?". Exceptions bubble up from the bottom not knowing where *or if* they will be handled. If an exception is caught by an intervening stack frame or fiber, then that intervening frame or fiber believes it knows how to handle that exception. As the error originator, it's basically none of my business where it gets handled.

But `break` and `return` and `throw` mean: "We're done now. I *do* know how to handle this: Exit *here* and unwind up to *there*."  All of these force execution to exit a *specific target* stack frame: `break` exits its block, `return` exits its method, and `throw` exits its matching `catch`. If they can't find their target above them in the stack then *that* is an exception to be raised. Likewise, cancellation is not (usually) an error to be handled; it's simply a command to return. Canceling one fiber implicitly necessitates all of its child fibers be canceled too.

Because of these differences, I don't believe that it makes sense for `Fiber#raise` to implicitly subvert the rules in https://bugs.ruby-lang.org/issues/17221 by following the chain and raising at the bottom. If you don't know what fiber is at the bottom of the resume stack, then it's none of your business sending data to it via `resume` or `transfer`, Similarly, it's none of your business raising an exception on it. That will lead to confusing bugs. On the other hand, `Fiber#cancel` sends neither a value nor an exception into the canceled fibers. The "cancel_reason" isn't for their consumption but for their `return_fiber`'s.

Reply:
> Is `return` just an unrescuable ruby exception? Is `break` just an unrescuable ruby exception? 

In my view, [yes](https://github.com/oracle/truffleruby/blob/master/src/main/java/org/truffleruby/language/control/LocalReturnException.java). And they behave very much like exceptions, e.g., they also run `ensure` blocks.
Now these two keywords also make it clear to which lexical scope around them they return, something which is not really feasible for stopping a Fiber (at least it's not lexical).
`throw`/`catch` is also just an exception, one that happens to not have a backtrace.
So I mean "exception" as a VM concept, not necessarily a Ruby Exception instance here.

> #3766 strictly adheres to the all of the protections in https://bugs.ruby-lang.org/issues/17221
> Because of these differences, I don't believe that it makes sense for `Fiber#raise` to implicitly subvert the rules in https://bugs.ruby-lang.org/issues/17221

Which one is it? Following the `resuming_fiber` chain is one of them but it can't be both. These guarantees clearly are not about the method name used. Either we consider raise/cancel to be "transferring control between Fibers" or we don't.

I get your point that `return/break/Fiber#cancel` cannot be `rescue`'d. But I'm unsure if it's a good idea.
Suppose a `resume` chain root->f1->f2->f3->f4. When you call `Fiber#cancel` on `f2` from say `f5` which was transferred to from `f4`, it will then cancel `f4`, then `f3`, then `f2`, then `f1`, then stop at the root?

But what if `f1` is some sort of Fiber pool or so, and nothing should escape it? Then Fiber#cancel just breaks that.

I think cancelling `f4` when doing `f2.cancel` feels a bit weird (i.e. you should know which Fiber is yielding and you want to stop), but I think it's exactly as weird whether using `cancel` or `raise`.

In your examples of https://bugs.ruby-lang.org/issues/17325#note-8, `a&.cancel rescue nil` feels like an anti-pattern. Ignoring exceptions is always a bad idea (it makes it a lot harder to debug anything).

Reply:
BTW, could you reply to my questions in https://bugs.ruby-lang.org/issues/17325#note-14? The more direct, the better.
Also some examples like the one I gave above with f1-f5 of differences between cancel and raise would be helpful on that ticket.
It seems you use "bottom of the stack" as current execution, which seems confusing if we consider a real-world stack and gravity (it seems the convention is the bottom doesn't move: https://en.wikipedia.org/wiki/Stack_(abstract_data_type)#Hardware_stack).

Reply:
> Suppose a `resume` chain root->f1->f2->f3->f4. When you call `Fiber#cancel` on `f2` from say `f5` which was transferred to from `f4`, it will then cancel `f4`, then `f3`, then `f2`, then `f1`, then stop at the root?

Why would that cancel `f1` or `root`?

```ruby
def foo
  loop do # loop 1
    stuff
    break :loop1
    other stuff
  end
  etc
  loop do # loop 2
    stuff
    break :loop2
    other stuff
  end
  etc
end
 ```

`break :loop1` doesn't cancel the entire method `foo`. Similarly, why would `f2.cancel :reason` cancel `f1`? `f2.cancel :reason` doesn't cause `f2` to terminate with an exception, but with the value `:reason`. It would simply return `:reason` to `f1`. I was going to add an exceptional cancel (behaving the same way on the inside, but causing the canceled fiber to terminate with an exception), because I think that'd be useful e.g. for timeouts. But I think the unexceptional cancel is much more common and much more useful.

I need to get off my computer ASAP now, but I'll write a couple of examples later. Very quickly though: `f2.cancel :reason` wouldn't (normally) return an exception. Since `f2` was resumed by `f1`, in the normal case `f2.cancel :reason` causes `f1`'s `f2.resume etc` to return `:reason` to `f1`. If it was called from `f5` then `f2.cancel :reason` will return... whatever is sent to `f5` the next time `f5.transfer etc` is called.

Did I misunderstand something in your example?

The abnormal cases  (i.e. `f2`'s `ensure` clauses never call `Fiber.yield` and don't raise an exception and `f2` isn't canceled a second time) are definitely worth interrogating more deeply. But I think that's part of `f2`'s contract with the other fibers it interacts with. As such, I don't want to impose any new rules on how `f2` transfers or yields. That's up to those fibers to continue to adhere to the rules they've already established between each other.

I guess I need to clean up [the tests](https://github.com/ruby/ruby/pull/3766/files#diff-09acabd1150d7f4c29c8173a9e18b35d7e6109b1e2cf0bbf61a5f63be558009bR233-R259) and [the docs](https://github.com/ruby/ruby/pull/3766/files#diff-f5e06781d15d3a6c16eef4cb49eebc310820bead3672ed84608c84958ae5da14R2339-R2362) so this process will be more obvious. I think I can make examples and documentation that make this fairly obvious, but those examples are a little bit too dense to be helpful.

> BTW, could you reply to my questions in https://bugs.ruby-lang.org/issues/17325#note-14? The more direct, the better.

Sorry. I'd intended to but got sidetracked. A big portion of the reply I'd written edited and placed into my comment above.

> Also some examples like the one I gave above with f1-f5 of differences between cancel and raise would be helpful on that ticket.

Yeah, sorry. Family and other obligations sidetracked me and I sometimes find it faster to write 1000 rambling words than 100 consise words. I was trying to edit it down to something useful.

> It seems you use "bottom of the stack" as current execution, which seems confusing if we consider a real-world stack and gravity (it seems the convention is the bottom doesn't move: https://en.wikipedia.org/wiki/Stack_(abstract_data_type)#Hardware_stack).

Sorry, yes. I often get the metaphors and directions of the stack mixed up and backwards. I do try to use less ambiguous words instead (current, last) but I often slip up. I think it's because it makes more sense to me when reading to read about where I've been to where I am (i.e. the stacktrace) from top to bottom, because that's the normal flow of reading. But you're right, the stack metaphor is usually spoken of in the bottom to top direction.

> In your examples of https://bugs.ruby-lang.org/issues/17325#note-8, `a&.cancel rescue nil` feels like an anti-pattern. Ignoring exceptions is always a bad idea (it makes it a lot harder to debug anything).

sorry, yes. the `cancel rescue nil` was just a quick toy example to gloss over "something happens here but I don't want to specify what, in order to keep the example short". In real code, exceptions would either need to be handled or unhandled (and thus raised to the next stack frame).

But in e.g. future or promise code, it's a very common pattern to snatch all exceptions (sometimes only `StandardError` or another class) and store them for later consumption. See e.g. [async's Task](https://github.com/socketry/async/blob/master/lib/async/task.rb#L264-L277) or [concurrent-ruby's IVar#set](https://github.com/ruby-concurrency/concurrent-ruby/blob/a5076bf9e977b290cce5a65b7b72810562bca295/lib/concurrent-ruby/concurrent/ivar.rb#L116-L121) or [SafeTaskExecutor#execute](https://github.com/ruby-concurrency/concurrent-ruby/blob/a5076bf9e977b290cce5a65b7b72810562bca295/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb#L18-L32). That's going to be less common in the case of fibers that are resumed. Generally `task` fibers like that will be either transferred to by the scheduler or resumed by the scheduler. Nevertheless, it's a common sort of pattern and it breaks `raise Exception, "cancel pls"` (but it doesn't break `break`).

Reply:
... Also, the picture examples in https://bugs.ruby-lang.org/issues/17221 are drawn as top to bottom. Execution stack direction is a metaphor that's very easy to flip upside-down! :laughing:

Reply:
> So I mean "exception" as a VM concept, not necessarily a Ruby Exception instance here.

Haha yeah, I thought you might respond with that. I'm aware that various forms of jumps back through the stack are generally handled through a similar mechanism at the VM level. My patch required me to do this: `ec->errinfo = (VALUE)THROW_DATA_NEW(reason, cfp, TAG_BREAK);` :slightly_smiling_face: I view "errinfo" here as a historically contingent term that stuck, because it clearly is used to hold non-err info. But that's probably not be how everyone else views it. :laughing:

My point is that from a *user's POV* and from an *API design POV*, they are different concepts. My `Fiber#cancel` patch isn't lexical like `break` or `return`--it's more similar to `throw`/`catch` in that respect. But all four of them share features: they all have a target that can be identified before they start unwinding the stack, and none of them can be rescued, and none of them are "errors" nor even exceptional (in the standard English sense of the word). All four of them are a statement of final certainty: we're done here and we know exactly where we want to go next. Clean up and exit now.

Raising exceptions has a very different logic and thus design: they unwind the stack with no specific target in mind. They are a statement of uncertainty. An error or anomaly or quirk or special case or deviation or--an exception--has occurred, and although we might describe it and classify it, we don't know (or can't know) how to handle it here in this stack frame. Send it to the the prior frame and maybe they'll have a guess about how to handle it. Or maybe not.

Although we can certainly attempt to tunnel semantic `throw` and `return` through an ruby exception (not talking about a VM exception :wink:), that seems sloppy to me, like the developer simply forgot about the existence of `throw` or `return` or `break`. Or sometimes we have an additional design constraint that means we really didn't know where to throw it--perhaps we're executing in an abnormal or dynamic context or perhaps we do want it to be optionally rescuable--that uncertainty *makes it* an exception and is why e.g. `Enumerator#next` must throw `StopIteration`.

It's not really anything to do with performance. And only partially to do with skipping `rescue` and `else`. It's primarily about certainty and intent: "Exit now." vs "I don't know how to handle this, maybe you do?."

-----

Anyway, back to _this_ PR. :laughing:  Another point re: the https://bugs.ruby-lang.org/issues/17221 protections.

I think of specific fibers as having an API for the messages that are passed back and forth to each other. The relationship between the "parent" (resumer) and "child" (resumed) fibers tends to be clearer than when transferring, but the same principle holds in both cases. If someone _other_ than the resumed child responds to the parent, maybe via a pre-3.0 `Fiber#transfer` or `#raise`, that probably breaks their API. Only the resumed child should pass messages back to the parent and only via `Fiber.yield` and only code that belongs to that API should be allowed to call `Fiber.yield`.

Basically my rule of thumb is: if your code wasn't part of the API that "owns" the current fiber, then you probably don't know who your return fiber is and you have no business calling `Fiber.yield` (nor pre-3.0 `#transfer`)--but `#resume` should be safe. And if your code doesn't "own" the target fiber and hasn't negotiated a mechanism for returning to the current fiber, then you have no business calling `#transfer` or `#resume`. If you violate those constraints, then you'll probably violate the API with the fiber on the receiving end and one or both fibers will receive messages they don't know how to handle. "Fiber#raise" is simply passing a special message, and so it should to adhere to these same rules.

Fortunately, the inverse is also true: if you are in a position to safely raise on a specific Fiber, then you probably *already know* which fiber you need to raise on and what they expect to receive. Or you can know without much effort. So there should be no need for sending an exception down the chain to an unknown fiber with an unknown API around resume and transfer.

This is exactly how many (most? all?) fiber-task libraries broke `Enumerator`. They would trigger `Fiber.yield` when they had no business calling `Fiber.yield`, from inside a fiber they didn't "own" e.g. an Enumerator fiber. And thus they broke the `Enumerator` API. This is easily fixed by resuming into a trampoline continuation fiber and _transferring_ from there into the scheduler:

```ruby
def suspending(&block)
  if yields_to_scheduler?
    suspender = -> waiter { Fiber.yield(Fiber.current, waiter) }
    scheduler_msg = block.call suspender
  elsif transfer_to_scheduler?
    suspender = -> waiter { scheduler_fiber.transfer(Fiber.current, waiter) }
    scheduler_msg = block.call suspender
  else
    # this is the trampoline continuation that most libraries don't have:
    # 1. resume trampoline; 2. transfer to scheduler; 3. transfer back; 4. trampoline terminates
    # This ensures that pre-3.0 Enumerator fibers are "undamaged" by the "transfer" bit.
    # This is unnecessary in 3.0, but it still works.
    #    with the caveat that the scheduler fiber must not be resuming
    scheduler_msg = Fiber.new do
      suspender = -> waiter { scheduler_fiber.transfer(Fiber.current, waiter) }
      block.call suspender
    end.resume
  end
  handle_scheduler_msg scheduler_msg
end

def io_wait(io, timeout=nil, mode=:read)
  suspending {|suspender| suspender.call io_waiter(io, timeout, mode) }
end
```

That trampoline continuation is what lets me ensure all of the fibers are playing by the rules and sending and receiving messages they understand. It's how I ensure my scheduler only transfers to fibers it "owns" (sorry I haven't gotten around to writing a PR to add a trampoline to `async` yet, it's a little bit more than just putting the trampoline into `Task.yield` because the scheduler also needs to be "transfer back to continuation" aware).

My point is, if some other fiber that has no part in these transfers and yields reaches in and raises an exception on a fiber they know nothing about, who knows what'll happen? You can't know how your exception will be received or handled if you don't even know who you're sending it to. It's very unpredictable. Flip side: with a little effort you probably *can* know the fiber you want to raise on and what it's waiting for.